File: day-22/crab_combat.py
import itertools
from collections import deque

input_file = 'day-22/input.txt'

def process_input(input):
  split = input.split('\n\n')
  player1 = deque([int(x) for x in split[0].splitlines()[1:]])
  player2 = deque([int(x) for x in split[1].splitlines()[1:]])
  return player1, player2

def play_round(player1, player2, part2 = False):
  previous_rounds = set() #TODO - tidy up cache
  p1_cache, p2_cache = set(), set()

  while player1 and player2:

    if part2:
      hash = (tuple(player1), tuple(player2))
      if hash in previous_rounds:
        return player1, []
      else:
        previous_rounds.add(hash)
  
    p1, p2 = player1.popleft(), player2.popleft()

    if part2 and (len(player1) >= p1 and len(player2) >= p2):
      player1_copy = list(itertools.islice(player1, 0, p1))
      player2_copy = list(itertools.islice(player2, 0, p2))

      rec1, rec2 = play_round(deque(player1_copy), deque(player2_copy), True)
      p1_win = len(rec2) == 0
    else:
      p1_win = p1 > p2

    if p1_win:
      player1 += [p1, p2]
    else:
      player2 += [p2, p1]

  return player1, player2

def get_score(queue):
  score = 0
  length = len(queue)
  for i, v in enumerate(queue):
    score += v * (length-i)
  return score

def play_game(input, part2 = False):
  player1, player2 = play_round(*process_input(input), part2)

  if len(player1) > 0:
    return get_score(player1)
  return get_score(player2)

def part1(input):
  return play_game(input)

def part2(input):
  return play_game(input, True)

if __name__ == "__main__":
  with open(input_file) as f:
    data = f.read()
  print("Part 1: ", part1(data))
  print("Part 2: ", part2(data))

File: day-22/crab_combat.go
package main

import (
	"fmt"
	"io/ioutil"
	"strconv"
	"strings"
)

const inputFile string = "./day-22/input.txt"

type deck []int

func (d deck) getScore() int {
	score := 0
	length := len(d)
	for i, v := range d {
		score += v * (length - i)
	}
	return score
}

func (d deck) copy(start int, end int) deck {
	return append(deck(nil), d[start:end]...)
}

func processDeck(input string) deck {
	var numbers deck
	for _, num := range strings.Split(strings.TrimSpace(input), "\n")[1:] {
		n, _ := strconv.Atoi(num)
		numbers = append(numbers, n)
	}
	return numbers
}

func processInput(input string) (deck, deck) {
	players := strings.Split(strings.TrimSpace(input), "\n\n")
	return processDeck(players[0]), processDeck(players[1])
}

type hash struct {
	hash1, hash2 int
}

func playRound(player1 deck, player2 deck, part2 bool) ([]int, []int) {
	previousRounds := make(map[hash]bool)

	for len(player1) > 0 && len(player2) > 0 {

		if part2 {
			roundHash := hash{player1.getScore(), player2.getScore()}
			if _, found := previousRounds[roundHash]; found {
				return player1, deck{}
			}
			previousRounds[roundHash] = true
		}

		var p1, p2 int
		p1, player1 = player1[0], player1[1:]
		p2, player2 = player2[0], player2[1:]

		var p1Win bool
		if part2 && len(player1) >= p1 && len(player2) >= p2 {
			copy1, copy2 := player1.copy(0, p1), player2.copy(0, p2)
			_, rec2 := playRound(copy1, copy2, true)
			p1Win = len(rec2) == 0
		} else {
			p1Win = p1 > p2
		}

		if p1Win {
			player1 = append(player1, p1, p2)
		} else {
			player2 = append(player2, p2, p1)
		}
	}
	return player1, player2
}

func playGame(input string, part2 bool) int {
	player1, player2 := processInput(input)

	player1, player2 = playRound(player1, player2, part2)

	if len(player1) > 0 {
		return player1.getScore()
	}
	return player2.getScore()
}

//Part1 - Solution to Part 1 of the puzzle
func Part1(input string) int {
	return playGame(input, false)
}

//Part2 - Solution to Part 2 of the puzzle
func Part2(input string) int {
	return playGame(input, true)
}

func main() {
	b, err := ioutil.ReadFile(inputFile)
	if err != nil {
		fmt.Print(err)
	}
	data := string(b)

	fmt.Printf("Part 1: %d\n", Part1(data))
	fmt.Printf("Part 2: %d\n", Part2(data))
}
